src/torrent.cpp

```diff
@@ -4074,6 +4076,13 @@ namespace {
 			{
 				ret = false;
 			}
+			else if (result.status == set_block_hash_result::result::piece_hash_failed && i == blocks_in_piece - 1)
+			{
+				// only if the *last* block causes the piece to fail, do we know
+				// it actually failed. Otherwise it might have been failing
+				// because of other, previously existing block hashes.
+				ret = false;
```

Comment:
without this, the extended test fails because even though the piece hash failed, the data would be marked as "finished" in the piece picker. The test now checks for that by ensuring that file progress is not 100% in the case where we don't have the full torrent.

I must admit I don't fully understand the case where *other* pieces than this one can be completed (in the loop above). I have a feeling that this check with `i` may need to be more sophisticated.

Reply:
after having spent some more time understanding this, I feel more confident that this is correct